planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+causing load there that we dont want. To be honest, I think this might not necessarily be the easiest problem to solve robustly.
+
+
+# Integrating with an event bus
+
+This is a bit of a special topic but I think that Im not alone in doing this so I think its worth putting my thoughts down on this topic as well. 
+We use MassTransit and RabbitMq and we have probably made _every single mistake that you can do_ publishing events on the bus. 
+
+We started out making the classic mistake of publishing our eventsourced events on the event bus. This was of course a bad move and everyone suffered.
+Eventsourced events != Eventbus events. We ended up building a separate message model where the messages that we post on the bus is enriched with additional
+information taken mainly from the aggregate. We plugged this into our homegrown integration library together with a hook for mapping between the events
+saved in Marten and the events published to the messagebus. This has served us very well.
+
+However now we end up with a different discussion. The dreaded 2PC. I saw somewhere (though I cant find it again now) someone suggesting using `SessionListener` 
+to publish to an event bus. _This way lies madness._ We ended up implementing the same algorithm that NServiceBus uses to "guarantee" eventual message 
+publishing after saving to the eventstore. This algorithm works, but its not the best performance-wise and it in theory requires infinite retries in order for
+the guarantee to actually be a guarantee. Right now we are where we are and reimplementing this part in our homegrown lib is not really on the table right
+now. But if I had to implement this again knowing what I do today I would 100% certainly implement publishing to the message bus as an async projection.
+
+Running projections off the message bus was also a thing that we tried doing for a while. It has some good parts, but also some bad parts. The main issue
+is of course that you suddenly end up having to handle things like out-of-order messages, duplicated messages, not being sure of where you are in the stream
+etc. Its quite a lot to keep in mind, but for some cases its quite useful. But complexity immediately rises, especially in writing the projections, and lets not
+even talk about event replay. I saw that you were discussing allowing projections from a message bus, and it would be a great feature to have but be prepared
+that in order to support the same type of projections as you can do today is gonna require quite a bit of infrastructure. But if you can pull it off, Id certainly be happy! ;)
+
+
+# Immutable objects
+
+Honestly, you knew this were coming. But hear me out.
+
+## Immutable events
+Events should be immutable, thats just a basic fact of eventsourcing. But due to serialization woes etc the common implementation of events are with 
+`public string Fruit { get; set; }` because thats just life in C# world. Now, this "gentleman immutability" "works" but it is...not great?
```

Comment:
@wastaz I'm using such pattern: https://github.com/oskardudycz/EventSourcing.NetCore/blob/master/Sample/Tickets/Tickets/Reservations/Events/TentativeReservationCreated.cs (I think that `[JsonConstructor]` is redundant).
This makes the event immutable from the code (of course, not from serialization) and works well with the Marten and Newtonsoft.